### PR TITLE
Support pointers of unmanaged type parameters in method type inference

### LINKED ISSUE #25654: Type inference doesn't work with a pointer of a generic unmanaged type
**Version Used**:  `dev15.7.x`

**Steps to Reproduce**:

```cs
class C {
    unsafe void M<T>(T* a) where T : unmanaged {
        var p = stackalloc T[10];
        M(p);
    }
}
```

**Expected Behavior**: No error

**Actual Behavior**:

`The type arguments for method 'C.M<T>(T*)' cannot be inferred from the usage. Try specifying the type arguments explicitly.`

/cc @OmarTawfik

### PR DESCRIPTION
**Pull request by OmarTawfik**
Fixes #25654 

@dotnet/roslyn-compiler for review.

### CODE REVIEW

**Comment by alrz** [comment]
How about `Microsoft.CSharp.RuntimeBinder.Semantics.MethodTypeInferrer`?

**Comment by OmarTawfik** [comment]
@alrz please feel free to open a bug there for that support. If not, I'll open one when I get some time.

**Comment by cston** [comment]
Should the following compile?
```
    static void F<T>(T* x, T* y) where T : unmanaged
    {
    }
    static void G<T>(void* x, T* y) where T : unmanaged
    {
        F(x, y);
    }
```

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs [hunk 1]
```diff
@@ -522,11 +522,10 @@ private void InferTypeArgsFirstPhase(Binder binder, ref HashSet<DiagnosticInfo>
             // SPEC: For each of the method arguments Ei:
             for (int arg = 0, length = this.NumberArgumentsToProcess; arg < length; arg++)
             {
-                var argument = _arguments[arg];
-
-                bool isExactInference = GetRefKind(arg) != RefKind.None;
-
+                BoundExpression argument = _arguments[arg];
                 TypeSymbol target = _formalParameterTypes[arg];
+                bool isExactInference = GetRefKind(arg).IsManagedReference() || argument.Type.IsPointerType() || target.IsPointerType();
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs]: Is there a test for successful inference when the argument type is not a pointer type? Perhaps:
```
static void F<T>(T* a, T b) where T : unmanaged
{
    F(null, b);
}
```

Comment (cston) [reply src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs]: Is the check for `argument.Type.IsPointerType()` necessary?

**Comment by OmarTawfik** [comment]
It shouldn't, as `void` will not be convertible to `T`. Why is that scenario different?

---
In reply to: [381843146](https://github.com/dotnet/roslyn/pull/26154#issuecomment-381843146) [](ancestors = 381843146)

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs (hunk 1 above)
Comment (OmarTawfik) [reply src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs]: No. I'll add one.

---
In reply to: [182085539](https://github.com/dotnet/roslyn/pull/26154#discussion_r182085539) [](ancestors = 182085539)

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs [hunk 2]
```diff
@@ -1598,6 +1603,17 @@ private bool ExactConstructedInference(TypeSymbol source, TypeSymbol target, ref
             return true;
         }
 
+        private bool ExactPointerInference(TypeSymbol source, TypeSymbol target, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        {
+            if (source is PointerTypeSymbol sourcePointer && target is PointerTypeSymbol targetPointer)
```
[the remark this answers is not in the document]
Comment (OmarTawfik) [reply src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs]: That would be faster if I don't need to cast, but in that case, I need to cast anyways to get the `PointedAtType`.

---
In reply to: [182085973](https://github.com/dotnet/roslyn/pull/26154#discussion_r182085973) [](ancestors = 182085973)

[the remark this answers is not in the document]
Comment (alrz) [reply src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs]: except for 99% of cases that we won't get a pointer here, kind check would be much faster than a type check.

**Comment by OmarTawfik** [comment]
@cston comments addressed
@alrz also opened https://github.com/dotnet/corefx/issues/29190
@dotnet/roslyn-compiler can I get a second review?

**Comment by OmarTawfik** [comment]
@jaredpar for approval.

**Comment by jaredpar** [comment]
approved

**Comment by OmarTawfik** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest please
https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_debug_vs-integration_prtest/707/
@dotnet-bot test windows_release_vs-integration_prtest please
https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_release_vs-integration_prtest/700/

**Comment by alrz** [comment]
This change is not compatible with the previous version 15.7, should that be documented?

**Pull request merged**